File: python/src/agent_framework/core/llm_summarization_compactor.py
from __future__ import annotations

from collections.abc import Awaitable, Callable

from agent_framework.core.context_compactor import estimate_tokens
from agent_framework.core.domain import Message

SummarizeMessages = Callable[[list[Message]], Awaitable[str]]


class LlmSummarizationCompactor:
    def __init__(self, summarize: SummarizeMessages, *, keep_recent: int = 4) -> None:
        self._summarize = summarize
        self._keep_recent = keep_recent

    async def compact(self, messages: list[Message], token_budget: int) -> list[Message]:
        total = sum(estimate_tokens(message) for message in messages)
        if total <= token_budget or len(messages) <= self._keep_recent + 1:
            return messages

        system = [message for message in messages if message.role == "system"]
        rest = [message for message in messages if message.role != "system"]
        head = rest[:1]
        tail = rest[-self._keep_recent :]
        middle = rest[len(head) : len(rest) - len(tail)]

        if not middle:
            return messages

        summary_text = await self._summarize(middle)
        summary = Message(role="system", content=f"[Summarized context]\n{summary_text}")
        compacted = [*system, *head, summary, *tail]

        if sum(estimate_tokens(message) for message in compacted) > token_budget and len(compacted) > 2:
            return [*system, *head, summary, *tail[-self._keep_recent :]]
        return compacted

File: typescript/packages/core/src/llm-summarization-compactor.ts
import type { ContextCompactor } from './types/contracts.js';
import type { Message } from './types/domain.js';
import { estimateTokens } from './context-compactor.js';

export type SummarizeMessages = (messages: Message[]) => Promise<string>;

export interface LlmSummarizationCompactorOptions {
  summarize: SummarizeMessages;
  keepRecent?: number;
}

export class LlmSummarizationCompactor implements ContextCompactor {
  private readonly keepRecent: number;

  constructor(private readonly options: LlmSummarizationCompactorOptions) {
    this.keepRecent = options.keepRecent ?? 4;
  }

  async compact(messages: Message[], tokenBudget: number): Promise<Message[]> {
    const total = messages.reduce((sum, message) => sum + estimateTokens(message), 0);
    if (total <= tokenBudget || messages.length <= this.keepRecent + 1) {
      return messages;
    }

    const system = messages.filter((message) => message.role === 'system');
    const rest = messages.filter((message) => message.role !== 'system');
    const head = rest.slice(0, 1);
    const tail = rest.slice(-this.keepRecent);
    const middle = rest.slice(head.length, rest.length - tail.length);

    if (middle.length === 0) {
      return messages;
    }

    const summaryText = await this.options.summarize(middle);
    const summary: Message = {
      role: 'system',
      content: `[Summarized context]\n${summaryText}`,
    };

    const compacted = [...system, ...head, summary, ...tail];
    const compactedTotal = compacted.reduce((sum, message) => sum + estimateTokens(message), 0);
    if (compactedTotal > tokenBudget && compacted.length > 2) {
      return [...system, ...head, summary, ...tail.slice(-this.keepRecent)];
    }
    return compacted;
  }
}
